
# MED 263: Final Project

---

## Submission deadline

Email a final project report pdf file to TA Andre Faustino (affaustino@ucsd.edu) by

```diff
- 11:59 pm Sunday March 18, 2018
```

## Final project presentation
Each student will do a 10-minute presentation on the last class day, Thursday March 8, 2018. This will help TA and instructors to understand more about your work.


## Sign-up your project
Start your final project by signing up on the [final project](http://bit.ly/med263students)


## Hands-on session material of your own

- You will create yet another MED263 hands-on session material plus homework-style questions and answers using software tools and data different from the ones used in the class.
- TA should be able to download the data following the instruction in your report.
- TA should be able to follow the instruction and code in your report to reproduce your results, also in the report.
- TA should be able to reproduce the answers to the questions provided in your report.
- You will submit a technical report in pdf format.
- The final project report will have sections of Introduction, Data, Methods/Results, and References. (Unlike typical journal, you will combine Methods and Results into one section.)
- Methods/Results section will have instruction, code, results (numbers/tables/figure), homework-style questions, and homework-style answers.
- There is no page limit.
- The example final project report (.pdf) can be found at [MED263 class GitHub](https://github.com/jihoonkim/MED263/raw/master/final/MED263_final_project_example.pdf).



## Recent data analysis workflows published in Nature Protocols 
Please feel free to replicate/modify published workflow with proper citation:

- [A high-throughput functional genomics workflow based on CRISPR/Cas9-mediated
 targeted mutagenesis in zebrafish](https://www.ncbi.nlm.nih.gov/pubmed/27809318)
- [Indel variant analysis of short-read sequencing data with Scalpel](https://www.ncbi.nlm.nih.gov/pubmed/27854363)
- [Highly multiplexed targeted DNA sequencing from single nuclei](https://www.ncbi.nlm.nih.gov/pubmed/26741407)
- [Methylation-assisted bisulfite sequencing to simultaneously map 5fC and 5caC 
on a genome-wide scale for DNA demethylation analysis](https://www.ncbi.nlm.nih.gov/pubmed/27281647)
- [MMRA MicroRNA Master Regulator Analysis](https://www.ncbi.nlm.nih.gov/pubmed/27305450)
- [The MaxQuant computational platform for mass spectrometry-based shotgun proteomics](https://www.ncbi.nlm.nih.gov/pubmed/27809316)
- [Using single nuclei for RNA-seq to capture the transcriptome of postmortem neurons](https://www.ncbi.nlm.nih.gov/pubmed/26890679)


## Data repositories

Visit following repositories to find the data:

- all different kinds of data in [DRYAD](http://datadryad.org)
- data published by [Broad Institute](https://www.broadinstitute.org/data-software-and-tools)
- gene expression data in [Gene Expression Omnibus](https://www.ncbi.nlm.nih.gov/geo)
- nucleotide data in [European Nucleotide Archive](http://www.ebi.ac.uk/ena)
- public genome data in [Personal Genome Project](http://www.personalgenomes.org)
- public genome data in [The International Genome Sample Resource](http://www.internationalgenome.org)
- proteomics data in [PRIDE PRoteomics IDEntification](http://www.ebi.ac.uk/pride/archive)

